test/jdk/java/net/httpclient/http2/IdleConnectionTimeoutTest.java

```diff
@@ -0,0 +1,161 @@
+        assertEquals(hresp.statusCode(), 200);
+        // Sleep for 4x the timeout value to ensure that it occurs
+        Thread.sleep(800);
+
+        request = hc.sendAsync(hreq, HttpResponse.BodyHandlers.ofString(UTF_8));
+        hresp = request.join();
+        assertEquals(hresp.statusCode(), 200);
+    }
+
+    /*
+        The opposite of testTimeoutFires(), if the same connection is used for both requests then the
+        idleConnectionTimeoutEvent did not occur.
+     */
+    @Test
+    public void testTimeoutDoesNotFire() throws InterruptedException {
+        expectTimeout = false;
+        System.setProperty(IDLE_CONN_PROPERTY, "800");
+        HttpClient hc = HttpClient.newBuilder().version(HTTP_2).build();
+        HttpRequest hreq = HttpRequest.newBuilder(noTimeoutUri).version(HTTP_2).GET().build();
+
+        CompletableFuture<HttpResponse<String>> request = hc.sendAsync(hreq, HttpResponse.BodyHandlers.ofString(UTF_8));
+        HttpResponse<String> hresp = request.join();
+        assertEquals(hresp.statusCode(), 200);
+        // Sleep for 1/8th of the timeout value to ensure it does not occur
+        Thread.sleep(100);
+
+        request = hc.sendAsync(hreq, HttpResponse.BodyHandlers.ofString(UTF_8));
+        hresp = request.join();
+        assertEquals(hresp.statusCode(), 200);
+    }
+
+    static class ServerTimeoutHandler implements Http2Handler {
+
+        InetSocketAddress remote;
+
+        @Override
+        public void handle(Http2TestExchange exchange) throws IOException {
+            if (remote == null) {
+                remote = exchange.getRemoteAddress();
+                exchange.sendResponseHeaders(200, 0);
+            } else if (!remote.equals(exchange.getRemoteAddress())) {
+                testLog.println("ServerTimeoutHandler: New Connection was used, idleConnectionTimeoutEvent fired."
+                        + " First remote: " + remote + ", Second Remote: " + exchange.getRemoteAddress());
+                exchange.sendResponseHeaders(200, 0);
+            } else {
+                exchange.sendResponseHeaders(400, 0);
+            }
+        }
+    }
+
+    static class ServerNoTimeoutHandler implements Http2Handler {
+
+        InetSocketAddress oldRemote;
+
+        @Override
+        public void handle(Http2TestExchange exchange) throws IOException {
+            InetSocketAddress newRemote = exchange.getRemoteAddress();
+            if (oldRemote == null) {
+                oldRemote = newRemote;
+                exchange.sendResponseHeaders(200, 0);
+            } else if (oldRemote.equals(newRemote)) {
+                testLog.println("ServerTimeoutHandler: Same Connection was used, idleConnectionTimeoutEvent did not fire."
+                        + " First remote: " + oldRemote + ", Second Remote: " + newRemote);
+                exchange.sendResponseHeaders(200, 0);
+            } else {
+                exchange.sendResponseHeaders(400, 0);
```

Comment:
Would it be useful to include a log message here which prints the expected (old) client address and the new client address, when it fails?